unixfs/README.md

```diff
@@ -37,6 +37,16 @@ message Data {
 
 	optional uint64 hashType = 5;
 	optional uint64 fanout = 6;
+	
+	/* file metadata */
+	optional uint32 mtime = 7;
+	optional uint32 atime = 8;
+	optional uint32 ctime = 9;
```

Comment:
@djdv pointed out that windows supports creation time.

Reply:
Not supporting fractional time is an oversight. Especially provided we might be stuck with this stop-gap for quite a while. I urge you to consider doing the following:
```
message Timespec {
       // Use of int64 is deliberate - negative epoch is super unlikely
       // Using sint64 would introduce zig-zag encoding ( harder to eyeball )
       // The varint representing the time of writing this is 5 bytes long
       // It will remain so until October 26, 3058 ( 34,359,738,367 )
       required int64 UnixEpoch = 1;

       // fixed32 ( always 4 bytes, no varint ), as nanosecs are often > 2^28
       // https://developers.google.com/protocol-buffers/docs/proto#scalar
       // NOTE: on the wire this value will be *little* endian
       optional fixed32 UnixEpochNsec = 2;
}
```

And using `Timespec` for `mtime`/`ctime`/etc ( `atime` is a really odd choice to carry but since it is optional it's not a big deal )

Reply:
I'd like to offer some perspective from the field in terms of what I have and have not, ever, needed when working with package management as well as containerland:

- mtime comes up sometimes.

- atime is completely useless.  It will often get set again in the course of... anything else you're doing.  The `atime` vs `relatime` vs `noatime` flags on common filesystems have also muddied the field beyond all recognition.  There is approximately one program in the universe known to look at atime, and it's something you've never heard of.  atime should be disregarded.  atime provides purely negative value.

- ctime is completely useless.  It can't even be expressed unless you're writing filesystem drivers.  While it so happens that we are, and we can, that doesn't mean it's a good idea, that it's useful, or that we should.  For example, if we were to support commands that just unpack/manifest files on an existing (e.g. ext4) filesystem, we would not be able to preserve ctime anyway.  Most other tooling exists within this limitation as well, and as a result, the world has evolved to effectively disregard ctime for any significant purposes; e.g., if there was a piece of software that would not work correctly if ctimes changed, it would fail to work in any container system, anything that uses tarballs, etc; effectively, it would be nonviable, and resultingly, such software tends not to exist.

I'd think long and hard before perpetuating these pieces of metadata just because we 'can'.  They're *remarkably* useless, unused, and in many contexts unus*able*.

- fractional timestamps I also have very mixed feelings on.  We can observe that most tar systems don't do them, in practice.  Yes, I know there's an extension for that.  Go check if your tar command actually does that by default.  Go stdlib tar doesn't.  Docker and such systems have elided fractional components for ages as well, since they've been using the go stdlib tar libraries.  By and large, nobody seems to have noticed or cared.  We can look at the rate of serious problems being reported about timestamp precision in e.g. docker and use it as a barometer for how much fractional timestamps matter.  I'd say it's "not much".

Reply:
I only want to store what is needed for the top level use cases (package managers) given this is not a permanent or even our ideal solution to these problems. My assumption is that they need something close to what `stat` gives them, which includes `atime` and `ctime`, but if nobody needs this we shouldn’t preserve this data.

My experience with fractional timestamps is that most FS API’s don’t give them to you by default and require an additional flag or option in order to see them, which didn’t meet the bar I had set, but again, if package managers need this then we can do it.

Reply:
Also, note that high precision 64bit timestamps **are** in the [current schema](https://github.com/ipfs/unixfs-v2/issues/29) for `unixfsv2`.

Reply:
>@djdv pointed out that windows supports creation time.

(sorry for the late reply)
Just to be clear this attribute also exists on a range of systems I'm aware of, and probably more.
Windows(CreationTime), Solaris (crtime), FreeBSD (st_birthtim), Darwin (st_birthtimespec).

Reply:
what’s the consensus here? keep all 3 or cut down to ctime and mtime?

Reply:
IMO a prerequisite for adding these info like 'ctime' to unixfs should be to have documented answers to:

- If a command like `ipfs get` is going to unpack content from unixfs into a tree of dirs and files on a regular (say, ext4, just to pick something) filesystem, and this were run by a non root user, what should the behavior be?
- Is there any use to ctime other than being effectively a random number generator in the input data, which users cannot for any practical purposes control?  Should commands to add files to unixfs have a default behavior to set it to a "zero" value?  If so, does that change the expected behavior of either 'unpack' or mount commands?

... because whoever implements code around this *will* discover these questions about five minutes after we merge this.

(Or maybe not, and we quietly introduce a defacto random number generator to our nice content addressable system without noticing, which would be... worse.)

Reply:
I should note I have fairly strong and (I'd like to think) fairly informed opinions about this, because I've previously [wrestled with it in the Timeless project](https://github.com/polydawn/go-timeless-api/blob/0ece408663edb9dbc6109ef8690f8425f0d8f5f4/filesetFilters.go)... and while I'm not here to tell you I had The Solution over there, I *will* tell you that underestimating the complexity of the normalizing filters system that's almost certain to end  up growing around this area would be a serious mistake.  It's a touchy thing and has been a source of many headaches and bugs for me.  Treat it cautiously and especially be wary of adding data if there's no plan for what its normalizers would look and act like.

Reply:
We could split this PR into two parts - merge the `mode` property (which seems relatively uncontentious and solves some real-world problems our users have today) and bottom out the `*time` related stuff separately?

Reply:
There’s two questions here that seem to be getting merged together:

1. Should *anyone ever* be allowed to add these time properties to the File.
2. How should IPFS *use* these properties, and should that behavior be optional or by default.

I’m inclined to follow @achingbrain’s advise and split it at this point. However, I’m going to keep `mtime` *in the schema* since it seems to be the least controversial and it gives us some room to experiment with `*time` issues in IPFS. Without it being in the schema we have no room to even experiment sufficiently.

We can work through the `*time` issues in the initial implementations and decide before we ship whether or not we actually want to use it in IPFS.

Reply:
>@djdv pointed out that windows supports creation time.

For some reason my previous comment is gone now. I had a list of operating systems with links to their properties...

This property is present in other systems as well, usually under the name `crtime`, or `birthtime`.